### PR TITLE
Magic Link Email

### PR DESCRIPTION
**Pull request by jimleeder123**
This is for the addition of...

- A new action - to create and then send the magic link email
- A new notification - to be sent within the action
- A placeholder route for the email to link to

There are also tests set up for everything. I wasn't sure if there should be a unit test for the notification or not; I couldn't see any on Tower so I have not done any.

I was also unsure whether it's best to have the signed url as an accessor method or function but have made this an accessor which seems to be better?

### CODE REVIEW

**Comment by jimleeder123** [comment]
[Magic link email](https://trello.com/c/JE6CmMPi/69-magic-link-email)

File: app/Support/Helpers.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+use Carbon\CarbonImmutable;
+
+class Helpers
+{
+    /**
+     * @param CarbonImmutable $datetime
+     * @return string
+     */
+    public static function wfgFormatDateTime(CarbonImmutable $datetime): string
```
Comment (gbuckingham89) [inline app/Support/Helpers.php:13]: Can we drop the "wfg" from the method name? 

Great job with the timezone handling.

Datetime formatting might be something we revisit / refactor in a later PR _(e.g. what happens when we want to format a date/time with seconds too?)_

File: app/Actions/MagicLinks/SendAction.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    /**
+     * @param User $user
+     * @param string $ipAddress
+     * @param string|null $intendedUrl
+     * @return void
+     */
+    public function handle(User $user, string $ipAddress, string $intendedUrl = null): void
+    {
+        $magicLink = new MagicLink();
+        $magicLink->user_id = $user->id;
```
Comment (gbuckingham89) [inline app/Actions/MagicLinks/SendAction.php:26]: Just as small thing, but it's best to use the relation methods where possible; 

```
$magicLink->user()->associate($user);
```

File: tests/Unit/Actions/MagicLinks/SendActionTest.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+         * @var MagicLink $magicLink
+         */
+        $magicLink = MagicLink::query()->first();
+
+        $this->assertEquals($user->id, $magicLink->user_id);
+        $this->assertNotNull($magicLink->requested_at);
+        $this->assertNotNull($magicLink->expires_at);
+        $this->assertNull($magicLink->authenticated_at);
+        $this->assertEquals($ipAddress, $magicLink->ip_requested_from);
+        $this->assertNull($magicLink->authenticated_at);
```
Comment (gbuckingham89) [inline tests/Unit/Actions/MagicLinks/SendActionTest.php:42]: Looks like a copy & paste typo - I think this should be `ip_authenticated_from`?

File: app/Models/MagicLink.php [hunk 4]
```diff
@@ -39,7 +42,9 @@ class MagicLink extends Model
      * @var array<string, string|class-string>
      */
     protected $casts = [
-        //
+        'requested_at' => 'date',
+        'expires_at' => 'date',
+        'authenticated_at' => 'date',
```
Comment (gbuckingham89) [inline app/Models/MagicLink.php:47]: To amen sure they're cast properly, these should all be `immutable_datetime`.

**Review by gbuckingham89 — CHANGES_REQUESTED** [review]
Nice work, especially the tests - they're spot on! No need for testing the notification itself really - probably a bit overkill.

Just a couple of small tweaks to make.

I think having the signed URL as an eloquent accessor is great - as it's sort of a property of the `MagicLink` model.

File: app/Support/Helpers.php (hunk 1 above)
Comment (jimleeder123) [reply app/Support/Helpers.php:13]: Done, thanks

File: app/Actions/MagicLinks/SendAction.php (hunk 2 above)
Comment (jimleeder123) [reply app/Actions/MagicLinks/SendAction.php:26]: Done, thanks

File: tests/Unit/Actions/MagicLinks/SendActionTest.php (hunk 3 above)
Comment (jimleeder123) [reply tests/Unit/Actions/MagicLinks/SendActionTest.php:42]: Done, thanks

File: app/Models/MagicLink.php (hunk 4 above)
Comment (jimleeder123) [reply app/Models/MagicLink.php:47]: Done, thanks. I was unsure if this would be effected by the immutable line the app service provider or not

**Review by gbuckingham89 — APPROVED** [review]
✅ Looks good!

**Pull request merged**